src/tests/ftest/util/yaml_utils.py

```diff
@@ -0,0 +1,469 @@
+    elif isinstance(obj, list):
+        # Recursively look for matches in each list entry
+        for item in obj:
+            add_matches(find_values(item, keys, None, val_type))
+
+    return matches
+
+
+class YamlUpdater():
+    """A class for updating placeholders in test yaml files."""
+
+    YAML_KEYS = OrderedDict(
+        [
+            ("test_servers", "_servers"),
+            ("test_clients", "_clients"),
+            ("bdev_list", "_storage"),
+            ("timeout", "_timeout"),
+            ("timeouts", "_timeout"),
+            ("clush_timeout", "_timeout"),
+            ("ior_timeout", "_timeout"),
+            ("job_manager_timeout", "_timeout"),
+            ("pattern_timeout", "_timeout"),
+            ("pool_query_timeout", "_timeout"),
+            ("rebuild_timeout", "_timeout"),
+            ("srv_timeout", "_timeout"),
+            ("storage_prepare_timeout", "_timeout"),
+            ("storage_format_timeout", "_timeout"),
+        ]
+    )
+
+    def __init__(self, logger, servers, clients, storage, timeout, override, verbose):
+        """Initialize a YamlUpdater object.
+
+        Args:
+            logger (logger): logger for the messages produced by this class
+            servers (NodeSet): set of hosts to use as servers in the test yaml
+            clients (NodeSet): set of hosts to use as clients in the test yaml
+            storage (str): updated storage information to apply to the test yaml
+            timeout (int): multiplier to apply to any timeouts specified in the test yaml
+            verbose (int): verbosity level
+        """
+        self.log = logger
+        self._servers = servers
+        self._clients = clients
+        self._storage = storage
+        self._timeout = timeout
+        self._override = override
+        self._verbose = verbose
+
+    @property
+    def placeholder_updates(self):
+        """Do placeholder replacements exist for updating the test yaml.
+
+        Returns:
+            bool: whether or not placeholder replacements exist
+        """
+        return self._servers or self._storage or self._timeout
+
+    def update(self, yaml_file, yaml_dir):
+        """Update the placeholders in the specified test yaml file.
+
+        Args:
+            yaml_file (str): test yaml file to update
+            yaml_dir (str): directory in which to write the updated test yaml file
+
+        Raises:
+            YamlException: if any placeholders are found without a replacement
+
+        Returns:
+            str: the updated test yaml file or None if no updates were made
+
+        """
+        self.log.debug("-" * 80)
+        replacements = self.get_replacements(yaml_file)
+        return self.apply_changes(yaml_file, yaml_dir, replacements)
+
+    def get_replacements(self, yaml_file):
+        """Determine the replacement values for the placeholders in the test yaml.
+
+        Args:
+            logger (logger): logger for the messages produced by this method
```

Comment:
Looks like we don't have `logger` parameter.

Reply:
Removed, thanks.